lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -913,57 +924,83 @@ selectSeqStatePendingIxs ssid =
 -- address discovery requires a root key to recognize addresses.
 instance PersistState (Rnd.RndState t) where
     insertState (wid, sl) st = do
-        rsid <- insert $ RndState wid sl
-            (W.getIndex (st ^. #accountIndex))
-            (st ^. #gen)
-        insertRndStateAddresses rsid False $ Rnd.addresses st
-        insertRndStateAddresses rsid True $ Rnd.pendingAddresses st
+        repsert (RndStateKey wid)
+            (RndState wid (W.getIndex (st ^. #accountIndex)))
+        insertRndStateAddresses wid sl (Rnd.addresses st)
+        cpid <- insert $ RndStateCheckpoint wid sl (st ^. #gen)
+        insertRndStatePending cpid (Rnd.pendingAddresses st)
 
     selectState (wid, sl) = runMaybeT $ do
         st <- MaybeT $ selectFirst
             [ RndStateWalletId ==. wid
+            ] []
+        stc <- MaybeT $ selectFirst
+            [ RndStateCheckpointWalletId ==. wid
             , RndStateCheckpointSlot ==. sl
             ] []
-        let (rsid, RndState _ _ accIx gen) = (entityKey st, entityVal st)
-        addresses <- lift $ selectRndStateAddresses rsid False
-        pendingAddresses <- lift $ selectRndStateAddresses rsid True
+        addresses <- lift $ selectRndStateAddresses wid sl
+        pendingAddresses <- lift $ selectRndStatePending (entityKey stc)
         rndKey <- lift $ selectRndStateKey wid
         pure $ Rnd.RndState
             { rndKey = rndKey
-            , accountIndex = W.Index accIx
+            , accountIndex = W.Index (rndStateAccountIndex (entityVal st))
             , addresses = addresses
             , pendingAddresses = pendingAddresses
-            , gen = gen
+            , gen = rndStateCheckpointGen (entityVal stc)
             }
 
     deleteState wid = do
-        rsid <- fmap entityKey <$> selectList [ RndStateWalletId ==. wid ] []
-        deleteWhere [ RndStateAddressRndStateId <-. rsid ]
+        cpid <- fmap entityKey <$> selectList [ RndStateCheckpointWalletId ==. wid ] []
+        deleteWhere [ RndStatePendingAddressCheckpointId <-. cpid ]
+        deleteWhere [ RndStateCheckpointWalletId ==. wid ]
+        deleteWhere [ RndStateAddressWalletId ==. wid ]
         deleteWhere [ RndStateWalletId ==. wid ]
 
 insertRndStateAddresses
-    :: RndStateId
-    -> Bool
+    :: W.WalletId
+    -> W.SlotId
+    -> Map (W.Index 'W.Hardened 'W.AccountK, W.Index 'W.Hardened 'W.AddressK) W.Address
```

Comment:
See [this other comment](https://github.com/input-output-hk/cardano-wallet/pull/672/files#r320179776).